src/tools/airtable.ts

```diff
@@ -0,0 +1,185 @@
+  name = "Airtable";
+  description =
+    "Can Query records from airtable tables in an airtable base. Use the action GET_SCHEMA to learn about the structure of the airtable base and QUERY_TABLE to request data from a table within the base.";
+
+  inputSchema() {
+    return z.object({
+      action: z
+        .enum(["GET_SCHEMA", "QUERY_TABLE"])
+        .describe(
+          "The action for the tool to take. GET_SCHEMA requests the airtable base table schema. QUERY_TABLE requests data from a selected table.",
+        ),
+      table: z
+        .string()
+        .optional()
+        .describe("The table ID to query when using the QUERY_TABLE action."),
+      fields: z
+        .array(z.string())
+        .optional()
+        .describe("The fields to return from the table query when using the QUERY_TABLE action."),
+      filterFormula: z
+        .string()
+        .optional()
+        .describe(
+          "The airtable filter formula to refine the query when using the QUERY_TABLE action.",
+        ),
+    });
+  }
+
+  public constructor(options: AirtableToolOptions) {
+    super(options);
+    this.options = options;
+    this.airtable = new Airtable({ endpointUrl: airtableApi, apiKey: options.apiToken }); // pragma: allowlist secret
+    this.base = this.airtable.base(options.baseId);
+  }
+
+  protected async _run(input: ToolInput<this>, _options?: BaseToolRunOptions) {
+    if (input.action === "GET_SCHEMA") {
+      const response = await this.getBaseTableSchema();
+      return new JSONToolOutput(response);
+    } else if (input.action === "QUERY_TABLE" && input.table != undefined) {
+      const response = await this.getTableContents(input.table, input.fields, input.filterFormula);
+      return new JSONToolOutput(response);
+    } else {
+      throw new Error("Invalid Action.");
+    }
+  }
+
+  /**
+   * Requests the Airtable Base Schema.
+   *
+   * The airtable module does not have this functionality, the tool requests directly
+   * from the HTTP API.
+   * @returns - SummarisedAirtableTableSchema
+   */
+  private async getBaseTableSchema(): Promise<SummarisedAirtableTableSchema[]> {
+    const atResponse = await fetch(`${airtableApi}/v0/meta/bases/${this.options.baseId}/tables`, {
+      headers: {
+        Authorization: `Bearer ${this.options.apiToken}`,
+      },
+    });
+
+    if (atResponse.ok) {
+      const schemadata = await atResponse.json();
+      const tableSchema: SummarisedAirtableTableSchema[] = [];
+      for (const table of schemadata.tables) {
+        const tableData: SummarisedAirtableTableSchema = {
+          name: table.name,
+          id: table.id,
+          fields: [],
+        };
+        if (table.description != undefined) {
+          tableData.description = table.description;
+        }
+        for (const field of table.fields) {
+          tableData.fields.push({ name: field.name, type: field.type });
+        }
+        tableSchema.push(tableData);
+      }
+
+      return tableSchema;
```

Comment:
This can be simplified.

```typescript
const tableSchema: SummarisedAirtableTableSchema[]  = schemadata.tables.map(table => ({
  name: table.name,
  id: table.id,
  description: table.description,
  fields: table.fields.map(field => ({ name: field.name, type: field.type }))
}))
```